kernel/comps/virtio/src/device/block/device.rs

```diff
@@ -279,17 +244,72 @@ impl DeviceInner {
         String::from_utf8(device_id).unwrap()
     }
 
-    /// Reads data from the block device, this function is blocking.
-    pub fn read(&self, sector_id: Sid, bufs: &[DmaStreamSlice]) {
-        // FIXME: Handling cases without id.
-        let id = self.id_allocator.lock().pop().unwrap() as usize;
+    /// Handles the irq issued from the device
+    fn do_handle_irq(&self) {
+        loop {
+            // Pops the complete request
+            let complete_request = {
+                let mut queue = self.queue.lock_irq_disabled();
+                let Ok((token, _)) = queue.pop_used() else {
+                    return;
+                };
+                self.submitted_requests.lock().remove(&token).unwrap()
+            };
+
+            // Handles the response
+            let id = complete_request.id as usize;
+            let resp_slice = DmaStreamSlice::new(&self.block_responses, id * RESP_SIZE, RESP_SIZE);
+            resp_slice.sync().unwrap();
+            let resp: BlockResp = resp_slice.read_val(0).unwrap();
+            self.id_allocator.lock().free(id);
+            match RespStatus::try_from(resp.status).unwrap() {
+                RespStatus::Ok => {}
+                // FIXME: Return an error instead of triggering a kernel panic
+                _ => panic!("io error in block device"),
+            };
+
+            // Synchronize DMA mapping if read from the device
+            if let BioType::Read = complete_request.bio_request.type_() {
+                complete_request
+                    .bufs
+                    .iter()
+                    .for_each(|(stream, offset, len)| {
+                        stream.sync(*offset..*offset + *len).unwrap();
+                    });
+            }
+
+            // Completes the bio request
+            complete_request.bio_request.bios().for_each(|bio| {
+                bio.complete(BioStatus::Complete);
+            });
+        }
+    }
+
+    /// Reads data from the device, this function is no-blocking.
+    fn do_read(&self, bio_request: BioRequest) {
+        let dma_streams: Vec<(DmaStream, usize, usize)> = bio_request
+            .bios()
+            .flat_map(|bio| {
+                bio.segments().iter().map(|segment| {
+                    let dma_stream =
+                        DmaStream::map(segment.pages().clone(), DmaDirection::FromDevice, false)
+                            .unwrap();
+                    (dma_stream, segment.offset(), segment.nbytes())
+                })
+            })
+            .collect();
+        let dma_slices: Vec<DmaStreamSlice> = dma_streams
+            .iter()
+            .map(|(stream, offset, len)| DmaStreamSlice::new(stream, *offset, *len))
+            .collect();
```

Comment:
The `do_write` function contains duplicated code segments. Refactoring this code into a standalone function would enhance readability and maintainability.

Reply:
Maybe  I can add a function to do the dma mapping

Reply:
Add `dma_stream_map`.